src/lang/uk.rs

```diff
@@ -0,0 +1,1233 @@
+                    MEGA_FLEXIONS[properties.number.index()][properties.declenation.index()]
+                };
+                words.push(
+                    format!("{}{}",
+                    MEGA_BASES[order - 1],
+                    mega_flexion
+                ));
+            }
+        }
+
+        Ok(words.join(" "))
+    }
+
+    fn float_to_cardinal(&self, num: BigFloat) -> Result<String, Num2Err> {
+        let whole = num.int();
+        let mut numerator = num.frac().abs();
+        if numerator.is_zero() {
+            return self.int_to_cardinal(whole);
+        }
+        let mut denominator = BigFloat::from(1);
+        while !numerator.frac().is_zero() { //this is BAD, we should use non-floating point format
+            numerator *= BigFloat::from(10);
+            denominator *= BigFloat::from(10);
+        }
+        let whole_properties = self.agreement_with_num(whole);
+        let whole_flexion = if whole_properties.number == GrammaticalNumber::Plural {
+            ADJECTIVE_HARD_FLEXIONS_PLURAL
+        } else {
+            ADJECTIVE_HARD_FLEXIONS_SINGULAR[Gender::Feminine.index()]
+        }[whole_properties.declenation.index()];
+        
+        let whole_lang = whole_properties.feminine();
+        let numerator_properties = self.agreement_with_num(numerator);
+        let numerator_lang = numerator_properties.feminine();
+        Ok(format!("{} ціл{} {} {}",
+            whole_lang.int_to_cardinal(whole)?,
+            whole_flexion,
+            numerator_lang.int_to_cardinal(numerator)?,
+            numerator_lang.to_ordinal(denominator)?,
+        ))
+    }
+
+    fn ordinal_flexion(&self, num: BigFloat) -> &'static str {
+        let tail = (num % BigFloat::from(100)).to_u64().unwrap();
+        let is_soft = tail%10==3 && tail!=13; //третій - the only soft adjective in numbers
+        let f= match (self.is_plural(), is_soft) {
+            (true, true) => &ADJECTIVE_SOFT_FLEXIONS_PLURAL,
+            (true, false) => &ADJECTIVE_HARD_FLEXIONS_PLURAL,
+            (false, true) => &ADJECTIVE_SOFT_FLEXIONS_SINGULAR[self.gender.index()],
+            (false, false) => &ADJECTIVE_HARD_FLEXIONS_SINGULAR[self.gender.index()],
+        };
+        f[self.declenation.index()]
+    }
+
+    fn ordinal_flexion_short(&self, num: BigFloat) -> &'static str {
+        let f = if self.is_plural() {
+            &ORDINAL_FLEXIONS_PLURAL_SHORT
+        } else{
+            let tail = (num % BigFloat::from(100)).to_u64().unwrap();
+            if tail%10==3 && tail!=13 { //третій - the only soft adjective in numbers        
+                &ORDINAL_SOFT_FLEXIONS_SINGULAR_SHORT[self.gender.index()]
+            } else {
+                &ORDINAL_HARD_FLEXIONS_SINGULAR_SHORT[self.gender.index()]
+            }
+        };
+        f[self.declenation.index()]
+    }
+}
+
+impl Language for Ukrainian {
+    fn to_cardinal(&self, num: BigFloat) -> Result<String, Num2Err> {
+        if num.is_inf_pos() {
```

Comment:
no test cases for `inf`, not for errors related (infiniteyear)

Reply:
Errors are filtered in to_words